driver-core/src/main/java/com/datastax/driver/core/ReplicationStategy.java

```diff
@@ -74,17 +74,25 @@ private SimpleStrategy(int replicationFactor) {
             this.replicationFactor = replicationFactor;
         }
 
-        Map<Token, Set<Host>> computeTokenToReplicaMap(Map<Token, Host> tokenToPrimary, List<Token> ring) {
+        Map<Token, Set<Host>> computeTokenToReplicaMap(String keyspaceName, Map<Token, Host> tokenToPrimary, List<Token> ring) {
 
             int rf = Math.min(replicationFactor, ring.size());
 
+            Map<Host, Set<Host>> hostSetMap = new HashMap<Host, Set<Host>>();
             Map<Token, Set<Host>> replicaMap = new HashMap<Token, Set<Host>>(tokenToPrimary.size());
             for (int i = 0; i < ring.size(); i++) {
-                // Consecutive sections of the ring can assigned to the same host
-                Set<Host> replicas = new LinkedHashSet<Host>();
-                for (int j = 0; j < ring.size() && replicas.size() < rf; j++)
-                    replicas.add(tokenToPrimary.get(getTokenWrapping(i+j, ring)));
-                replicaMap.put(ring.get(i), ImmutableSet.copyOf(replicas));
+                // use a single replicas set for each host for this strategy
+                Host primaryHost = tokenToPrimary.get(getTokenWrapping(i, ring));
+                Set<Host> replicas = hostSetMap.get(primaryHost);
+                if (replicas == null) {
+                    // Consecutive sections of the ring can assigned to the same host
+                    replicas = new LinkedHashSet<Host>();
+                    replicas.add(primaryHost);
+                    for (int j = 1; j < ring.size() && replicas.size() < rf; j++)
+                        replicas.add(tokenToPrimary.get(getTokenWrapping(i + j, ring)));
+                    hostSetMap.put(primaryHost, replicas = ImmutableSet.copyOf(replicas));
+                }
+                replicaMap.put(ring.get(i), replicas);
```

Comment:
We can't memoize `replicas` based on `primaryHost`. What determines the contents of `replicas` is the section of the ring after token `i`; but multiple tokens can be assigned to the same host, so for different start positions you might get the same `primaryHost`, but what you encounter next on the ring will not be the same. That's why the tests fail.